### PR TITLE
Delete before write

### PR DESCRIPTION
**Pull request by jbornemann**
Changes Made (Separate commits) : 

- Delete before write functionality
- Workflow_OFF/deleteBeforeWrite steps are executed concurrently (as no dependency on thread shared inputstream) 
- Updated README to reflect delete before write functionality 
- Renamed "PreprocessTasklet, Preprocess*, etc" to "NamespaceSyncTasklet, Namespace*, etc" as there are more than one preprocessors now, so the semantics became confusing.  
 
For https://github.com/TWCable/grabbit/issues/20

Please review @jdigger @sikarwarvimal @sagarsane

### CODE REVIEW

**Comment by sikarwarvimal** [comment]
@jbornemann is this ready for review and testing?

**Comment by sikarwarvimal** [comment]
It is a old branch but it seems you are still working on this?

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+@Slf4j
+@CompileStatic
+class DeleteBeforeWriteTasklet implements Tasklet {
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Can we get a javadoc for the class?

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+    private String jobPath
+
+    private Collection<String> excludePaths
+
+    private SlingRepository slingRepository
+
+    void setExcludePaths(String paths) {
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Should add javadocs explaining the format that this is expected to be in, referencing the fact that `ClientBatchJob` is what populates the JobParameters with this as a single string.

Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Given that I am anal when it comes to immutability and such: This Tasklet is meant to be immutable, so these values should all be passed into the constructor, not setters.
```xml
<constructor-arg name="jobPath" value="#{jobParameters[path]}"/>
```

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        }
+        //Recurse on each
+        nodeAndExcludePaths.each {
+            deletePartialTree(rootNode.getNode(it.nodeName), it.excludePaths)
+        }
+    }
+
+    /**
+     * If we have a path for a node such as 'foo/bar/doo'
+     * node is 'foo', and the excludePaths is '/bar/doo'
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: There should be a "then" joining those two lines.

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+     */
+    static class NodeAndExcludePaths {
+
+        String nodeName
+        Collection<String> excludePaths
+
+        static NodeAndExcludePaths fromPath(final String path) {
+            //Remove leading '/'
+            final String thisPath = path.replaceFirst('/', '')
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Actually, given the example in the javadoc, this would mess things up.

"foo/bar/doo" -> {"foobar", "/doo"}

Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Safer:
```groovy
final thisPath = path.startsWith('/') ? path.substring(1) : path
```

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+    }
+
+    /**
+     * If we have a path for a node such as 'foo/bar/doo'
+     * node is 'foo', and the excludePaths is '/bar/doo'
+     */
+    static class NodeAndExcludePaths {
+
+        String nodeName
+        Collection<String> excludePaths
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Either make this immutable (which I see from line 112 it is not) or explicitly declare that it's expected to mutate.

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    void setExcludePaths(String paths) {
+        final thisPathsString = paths?.trim()
+        if(!thisPathsString) {
+            excludePaths = []
+            return
+        }
+        Collection<String> theseExcludePaths = thisPathsString.split("\\*") as Collection
+        excludePaths = theseExcludePaths.collect { String excludePath ->
+            return (excludePath[0] != '/') ? "/${excludePath}".toString() : excludePath
+        }
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: This is better replaced by
```groovy
def jobPathSlash = jobPath + '/'
if (excludePaths.any {!it.startsWith(jobPathSlash)})
    throw new IllegalArgumentException("Not all exclude paths start with \"${jobPathSlash}\": ${excludePaths}")
```
because then (assuming things have been changed over to constructor form and the valid format of `jobPath` has been assured) the real purpose of this check (that the exclude paths are all absolute) will be met, as well as guaranteeing that they are all relative to the jobPath.

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    private void deletePartialTree(Node rootNode, Collection<String> relativeExcludePaths) {
+        //Base case.  If a root node does not have any exclude paths under it, it must have been excluded
+        if(!relativeExcludePaths) {
+            return
+        }
+        //Compute the current tree level excluded nodes, and compute the remaining paths to traverse
+        Collection<NodeAndExcludePaths> nodeAndExcludePaths = relativeExcludePaths.inject([] as Collection<NodeAndExcludePaths>) { def acc, def thisPath ->
+            final thisNodeAndExcludePath = NodeAndExcludePaths.fromPath(thisPath)
+            //If we have already created a nodeAndExcludePath for this node name, just add to it's exclude paths
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Grammar Nazi: "just add it to it is paths?" :speak_no_evil:

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy (hunk 6 above)
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Even better, as I continue to think through this, would be
```groovy
def theExcludePaths = thisPathsString.split("\\*") as Collection<String>
def jobPathSlash = jobPath + '/'
if (theExcludePaths.any {!it.startsWith(jobPathSlash)}) throw new IllegalStateException("Not all exclude paths start with \"${jobPath}\": ${theExcludePaths}")
this.excludePaths = theExcludePaths.collect { it.substring(jobPathSlash.length()) }
```
then you can always treat the exclude paths as relative (since later they are converted that way anyway) **_and_** you've already validated that they are legit for the job path.

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        else {
+            deletePartialTree(session)
+        }
+        session.save()
+        return RepeatStatus.FINISHED
+    }
+
+
+    private void deletePartialTree(final Session session) {
+        final relativeExcludePaths = excludePaths.collect { it - jobPath }
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: This is no longer needed if `excludePaths` is set to be relative when set.

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+            //Remove leading '/'
+            final String thisPath = path.replaceFirst('/', '')
+            //We need to find the current node, and the remaining path e.g 'foo/bar/doo', node: foo, remaining path: /bar/doo
+            final slashIndex = thisPath.indexOf('/')
+            //There is no remaining path
+            if(slashIndex == -1) {
+                return new NodeAndExcludePaths(nodeName: thisPath, excludePaths: [])
+            }
+            final remainingPath = thisPath.substring(slashIndex)
+            final node = thisPath - remainingPath
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: More efficient than the `minus(..)`
```groovy
final node = thisPath.substring(0, slashIndex)
```
also a bit clearer as to the intent and less likely to get caught on subtle bugs like `remainingPath` happening to be a substring in what is supposed to be `node`

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+        static NodeAndExcludePaths fromPath(final String path) {
+            //Remove leading '/'
+            final String thisPath = path.replaceFirst('/', '')
+            //We need to find the current node, and the remaining path e.g 'foo/bar/doo', node: foo, remaining path: /bar/doo
+            final slashIndex = thisPath.indexOf('/')
+            //There is no remaining path
+            if(slashIndex == -1) {
+                return new NodeAndExcludePaths(nodeName: thisPath, excludePaths: [])
+            }
+            final remainingPath = thisPath.substring(slashIndex)
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: If you add 1 to the index then you're guaranteed to get a true relative path for the remaining path
```groovy
final remainingPath = thisPath.substring(slashIndex+1)
```

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy (hunk 4 above)
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: If `remainingPath` below, and `excludePaths` above are both guaranteed to be relative, then this whole piece about removing a leading `/` is no longer needed.

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+        Collection<String> theseExcludePaths = thisPathsString.split("\\*") as Collection
+        excludePaths = theseExcludePaths.collect { String excludePath ->
+            return (excludePath[0] != '/') ? "/${excludePath}".toString() : excludePath
+        }
+    }
+
+
+    void setJobPath(String jobPath) {
+        String thisJobPath = jobPath.trim()
+        thisJobPath = thisJobPath[0] != '/' ? "/${thisJobPath}" : thisJobPath
```
Comment (jdigger) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: May want to add
```groovy
thisJobPath = (thisJobPath[-1] != '/') ? thisJobPath : thisJobPath[0..-2]
```
to guard against `/foo/` being passed in.

**Comment by jdigger** [comment]
What should happen if a job path of `/fooble` is passed in but there's no path in the JCR by that name?

What should happen if a job path of `/foo` is passed, which exists, but an exclude path of `/foo/bar` is given but there's no path in the JCR by that name?

**Comment by jdigger** [comment]
FWIW, I love the use of comments in the various sections...

**Comment by jbornemann** [comment]
@jdigger ready for another round

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+     * @param contribution mutable state to be passed back to update the current
+     * step execution
+     * @param chunkContext attributes shared between invocations but not between
+     * restarts
+     * @return an {@link RepeatStatus} indicating whether processing is
+     * continuable.
+     */
+    @Override
+    RepeatStatus execute(StepContribution contribution, ChunkContext chunkContext) throws Exception {
+        final Session session = JcrUtil.getSession(slingRepository, "admin")
```
Comment (sagarsane) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Do we need a new session here? Can we not use the one persisted in `ThreadLocal`? (ClientBatchJobContext) ?

File: grabbit/src/main/resources/META-INF/spring/client-batch-job.xml [hunk 13]
```diff
@@ -32,16 +32,35 @@
...
     <batch:job id="clientJob" xmlns="http://www.springframework.org/schema/batch" job-repository="clientJobRepository">
-        <batch:step id="clientPreprocessor" next="clientWorkflowOff">
-            <batch:tasklet ref="clientPreprocessTasklet" transaction-manager="clientTransactionManager"/>
-        </batch:step>
-        <batch:step id="clientWorkflowOff" next="clientJcrNodes">
-            <batch:tasklet ref="clientWorkflowOffTasklet" transaction-manager="clientTransactionManager"/>
+        <!--  Do any presync tasks concurrently that we can, e.g turning off workflows -->
+        <batch:split id="preSyncFlow" task-executor="asyncTaskExecutor" next="clientNamespaceSync">
+            <batch:flow>
+                <batch:step id="clientWorkflowOff">
```
Comment (sagarsane) [inline grabbit/src/main/resources/META-INF/spring/client-batch-job.xml]: `clientWofkflowOff` is VERY lightweight step. Having said that, even if it is light weight, it actually does manage concurrency internally (so there is already multi-threading handling involved in the actual step and on that we are spawning different thread for the step it self).

Do we know this split of steps is going to give us tangible benefits?

If not, then I vote we keep it simple and have the steps be sequential. If there are no tangible benefits, I think this way will be harder to maintain (as you know we already have 1 Thread Per Path in the config)

cc: @jdigger

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+            deletePartialTree(rootNode, relativeExcludePaths)
+        }
+
+        session.save()
+        return RepeatStatus.FINISHED
+    }
+
+
+    /**
+     * Called recursively taking a rootNode, and deleting any nodes adjacent except adjacent corresponding nodes form relativeExcludePaths
```
Comment (sagarsane) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: I feel `and deleting any nodes adjacent except adjacent corresponding nodes form relativeExcludePaths` is a bit confusing and could be clearer :-)

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 15]
```diff
@@ -0,0 +1,214 @@
...
+
+        //If we don't have any exclude paths, we don't have to worry about only clearing particular subtrees
+        if(!relativeExcludePaths || relativeExcludePaths.empty) {
+            rootNode.remove()
+        }
+        else {
+            deletePartialTree(rootNode, relativeExcludePaths)
+        }
+
+        session.save()
```
Comment (sagarsane) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: This is just a suggestion (and open for discussion too :) )but you could implement a `DeleteBeforeWriteStepExecutionListener` that prepares (in this case, I guess there's nothing much to prepare; but you can do anything in it) and cleans up after the step (in this case would be `session.save()`)

May be it will be cleaner? :) (example : https://github.com/TWCable/grabbit/blob/master/grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/jcrnodes/JcrNodesStepExecutionListener.groovy )

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy [hunk 16]
```diff
@@ -0,0 +1,214 @@
...
+            rootNode = session.getNode(jobPath)
+        }
+        catch(PathNotFoundException ex) {
+            log.warn "deleteBeforeWrite was enabled for jobPath ${jobPath} but job path does not exist on client!"
+            return RepeatStatus.FINISHED
+        }
+
+        //If we don't have any exclude paths, we don't have to worry about only clearing particular subtrees
+        if(!relativeExcludePaths || relativeExcludePaths.empty) {
+            rootNode.remove()
```
Comment (sagarsane) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Wondering if there is a preference / difference between `node.remove()` and `session.removeItem(/path)` ?

File: grabbit/src/main/resources/META-INF/spring/client-batch-job.xml (hunk 13 above)
Comment (jbornemann) [inline grabbit/src/main/resources/META-INF/spring/client-batch-job.xml]: Performance benefits are marginal at this point; but any small wins are great, especially since the complexity is managed by Spring Batch.  That said it's more to establish the pattern, that we should be doing independent tasks concurrently.

I hope to move namespaceSync into preFlowSync eventually, but for the sake of keeping the PR small...This will make this change a bit more significant.

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy (hunk 14 above)
Comment (jbornemann) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Oops, typo.  

Open to suggestions :-). I'll try to make it more clear

File: grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy (hunk 12 above)
Comment (jbornemann) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: This isn't managed by SimpleSynchronousTaskExecutioner (default) like clientJcrNodes, and namespaceSync are.  It's managed by it's async counterpart, so we can't rely on our ThreadLocal "hack".  

The session really only needs to be shared in clientJcrNodes in my opinion.  You probably know best with your research, but wondering if there is a better way to manage a session within clientJcrNodes ( Setting as a task executioner property maybe?) It's sort of an unfortunate leaky abstraction we have to rely on

Comment (sagarsane) [inline grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy]: Yes I know using ThreadLocal is kind of a last resort general but there were several things that made us use ThreadLocal instead of anything else.

1. The design was simpler to have 1 session per path and we had to share the incoming input stream between steps anyways (not just `clientJcrNodes`)
2. Originally, I was trying to use Spring Batch's `ExecutionContext` to store the session (and the input stream). But, Spring Batch discourages placing large objects in its ExecutionContext, which is understandable since ExecutionContext is one of the objects persisted in the `Spring Batch's Job Repository`.. which for our case, we have our `JcrJobRepository` layer.
3. Hence, ThreadLocal was the last resort and if done properly, we know that it would always work and as mentioned above, we had to share the `inputStream` amongst the steps anyways.

File: grabbit/src/main/resources/META-INF/spring/client-batch-job.xml (hunk 13 above)
Comment (sagarsane) [inline grabbit/src/main/resources/META-INF/spring/client-batch-job.xml]: I do think that using the feature will make sense in some cases, but I still don't think that parallelizing `clientWorkflowOffTasklet` and `deleteBeforeWriteTasklet` makes sense (and may be its not safe either). 

For example : I think its perfectly reasonable to have a workflow that will trigger when a DAM Asset is removed (may be to remove `custom renditions` for that Asset or something).

For such a scenario, if we parallellize the two tasks above, it is possible that the nodes are removed before the corresponding workflows are turned off, which will result in undesirable effects.

I could however see `nameSpaceSync` and `clientWorkflowOff` tasks being parallelized. I think that will be safe, but the current structure I do not think is safe in all the cases.

What do others think @jdigger @jazeren1 @sikarwarvimal Paul?

Comment (jbornemann) [inline grabbit/src/main/resources/META-INF/spring/client-batch-job.xml]: Although unlikely, I could see a case for that.  I'll revert it back to synchronous, and we can explore making nameSpaceSync, and deleteBeforeWrite concurrent at a later date.

Comment (jbornemann) [inline grabbit/src/main/resources/META-INF/spring/client-batch-job.xml]: https://github.com/TWCable/grabbit/issues/54

Comment (sagarsane) [inline grabbit/src/main/resources/META-INF/spring/client-batch-job.xml]: Cool. The revert will be done in a separate PR I presume?
